README.md
# canadacovidmetrics

This is a Python package that provides key metrics regarding COVID-19 situation in Canada across provinces using the [OpenCovid API](https://opencovid.ca/api/).

## Summary

This package allows users to obtain key metrics on COVID-19 situation in Canada at national or provincial level for a specific time period. The 4 functions will return key metrics, including total cumulative cases, total cumulative deaths, total cumulative recovered cases and total cumulative vaccine completion, using data from [OpenCovid API](https://opencovid.ca/api/). The users may use the key metrics to conduct further analyses on COVID-19 situation in Canada.

## Functions

There are 4 functions in this package:

-   `get_cases` Query total cumulative cases with ability to specify province and date range of returned data.

-   `get_deaths` Query total cumulative deaths with ability to specify province and date range of returned data.

-   `get_recoveries` Query total cumulative recovered cases with ability to specify province and date range of returned data.

-   `get_vaccinations` Query total cumulative vaccine completion with ability to specify province and date range of returned data.

## Installation

```bash
$ pip install "git+https://github.com/UBC-MDS/canadacovidmetrics"
```

## Usage & Examples

### Total number of deaths over past week by province

```python
from canadacovidmetrics import canadacovidmetrics as ccm
import datetime as dt
deaths_last_week = ccm.get_deaths(after=str(dt.date.today() - dt.timedelta(days=7)))
deaths_last_week.groupby('province').sum('deaths').plot.barh(y='deaths', title='Deaths by province in past week');
```

### National vaccination completion in 2021

```python
from canadacovidmetrics import canadacovidmetrics as ccm
vaccines_2021 = ccm.get_vaccinations(loc='canada', after='2021-01-01', before='2021-12-31')
vaccines_2021.plot('date_vaccine_completed', 'cumulative_cvaccine', title='Cumulative national vaccinations');
```

### Daily new case count by province

```python
from canadacovidmetrics import canadacovidmetrics as ccm
cases_data = ccm.get_cases().set_index('date_report')
cases_data.groupby('province')['cases'].plot(legend=True, figsize=(10,6), title='Number of reported cases by day by province');
```

## Python ecosystem

There are several packages for easy access to COVID-19 key metrics or data using different APIs, examples include
- [covid](https://github.com/nf1s/covid) using [John Hopkins University API](https://coronavirus.jhu.edu/about/)
- [COVID19Py](https://github.com/Kamaropoulos/COVID19Py) using [Coronavirus Tracker API](https://github.com/ExpDev07/coronavirus-tracker-api)
- [covid19pyclient](https://github.com/NiklasTiede/covid19pyclient) using [RKI API](https://github.com/marlon360/rki-covid-api)

To our knowledge, there is no similar package using [OpenCovid API](https://opencovid.ca/api/) in the Python ecosystem.

## Documentation

Documentation canadacovidmetrics can be found at [Read the Docs](https://canadacovidmetrics.readthedocs.io/)

## Contributors

-   Adam Morphy (@adammorphy)
-   Brandon Lam (@ming0701)
-   Lakshmi Santosha Valli Akella (@valli180)
-   Luke Collins (@LukeAC)

We welcome and recognize all contributions. Please find the guide for contribution in [Contributing Document](https://github.com/UBC-MDS/canadacovidmetrics/blob/main/CONTRIBUTING.md).

## License

`canadacovidmetrics` was created by the Contributors. It is licensed under the terms of the MIT license.

## Credits

`canadacovidmetrics` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.2.5 (2022-02-04)
### Documentation
* Updated to Creative Commons License ([`1f8138a`](https://github.com/UBC-MDS/canadacovidmetrics/commit/1f8138a8253bd09a1b6c82c87c042c5d0a3c9595))

## v0.2.4 (2022-02-04)
### Fix
* Update conf file to resolve examples module import issue. ([`4b79b4e`](https://github.com/UBC-MDS/canadacovidmetrics/commit/4b79b4ee8bff81d351e01a7ed50d5cb03bb27939))

## v0.2.3 (2022-01-29)


## v0.2.2 (2022-01-29)


## v0.2.1 (2022-01-29)


## v0.2.0 (2022-01-28)
### Feature
* Update function names ([`fe4f0ca`](https://github.com/UBC-MDS/canadacovidmetrics/commit/fe4f0cab50b8e690643e77f1a7ba1c904f0cd2e6))
* Update function names ([`c7ca396`](https://github.com/UBC-MDS/canadacovidmetrics/commit/c7ca396a4f225c5dbf740ed74a79362b56a279ce))
* Update example.ipynb ([`1b4e731`](https://github.com/UBC-MDS/canadacovidmetrics/commit/1b4e731a6f50e91f9bec7a6fef93b390d672508e))
* Update README.md with Usage and Examples section ([`ecb194c`](https://github.com/UBC-MDS/canadacovidmetrics/commit/ecb194ce9d4f911dda1c712dd2bb3217bfaed459))
* Change default datetype arg in all funcs. ([`96dbcae`](https://github.com/UBC-MDS/canadacovidmetrics/commit/96dbcae5eff601bc7248eedc728cd4542125ebdb))
* Datetype_type parameter added with tests ([`b92be17`](https://github.com/UBC-MDS/canadacovidmetrics/commit/b92be17205c3d03e438b8c930aeb34f4b39f4e1f))
* Add custom assert false messages in tests ([`1ea5aef`](https://github.com/UBC-MDS/canadacovidmetrics/commit/1ea5aef95a03eb0b2ea5c74da31ec284dbbd8256))
* Total_cumulative_vaccine_completion function implimented ([`2c4d241`](https://github.com/UBC-MDS/canadacovidmetrics/commit/2c4d2416f10a9a4cab7bf2d398b09bcf430aa40d))
* Create test for total_cumulative_recovered_cases ([`6f31b5d`](https://github.com/UBC-MDS/canadacovidmetrics/commit/6f31b5d43cf6e6c5c5d00288e4586b1a69dd3d0a))
* Create test for loc_format_check ([`e7cee7b`](https://github.com/UBC-MDS/canadacovidmetrics/commit/e7cee7bcacb1dd4ad7430b6bb45e2c2b681436f5))
* Create test for date_format_check ([`f29ff3f`](https://github.com/UBC-MDS/canadacovidmetrics/commit/f29ff3f3ca9a16501b84c044f9b0e967666f4845))
* Added location and date format error handlers ([`0af5b77`](https://github.com/UBC-MDS/canadacovidmetrics/commit/0af5b7757d93ca38cdfa6a774b7389ef111ac031))
* Add cumulative cases recovered function body ([`1fafb53`](https://github.com/UBC-MDS/canadacovidmetrics/commit/1fafb5332987578eb63e941c7e2f1efd3afac165))
* Add function docstrings ([`7915cff`](https://github.com/UBC-MDS/canadacovidmetrics/commit/7915cff6cd4d3267bff805e3a116cc97df2d8517))
* Add function shells ([`e20bf21`](https://github.com/UBC-MDS/canadacovidmetrics/commit/e20bf21a8810bc42989175f2f83104631f3d3437))

### Fix
* Revert pkg dependencies. Add pkg dev dependencies for sphinx ([`7fb886c`](https://github.com/UBC-MDS/canadacovidmetrics/commit/7fb886c5fb0bd74437448006f41bef59d32ce1ae))
* Add sphinx to dev dependencies ([`e7dd9f7`](https://github.com/UBC-MDS/canadacovidmetrics/commit/e7dd9f720ef80c65fa78a67f523eb2721fadffed))
* Set lowercase syntax for all instances of module/package name ([`b3f0fe0`](https://github.com/UBC-MDS/canadacovidmetrics/commit/b3f0fe08acd0446e010d16dd7681d48f87eb35f5))
* Update import principal package statement ([`d7a91da`](https://github.com/UBC-MDS/canadacovidmetrics/commit/d7a91daad1cc104ff4ae12ecdd03b2c5fa37363d))
* Set datetype to False for vacc_comp tests ([`73f5bf5`](https://github.com/UBC-MDS/canadacovidmetrics/commit/73f5bf52357e4c32e5aa8c4cc71e0a70b85b9def))
* Datetime correction ([`beb1af5`](https://github.com/UBC-MDS/canadacovidmetrics/commit/beb1af5e5fa5dc74a3f904fd6485494f82f54a42))
* Datetime use correction ([`341f5f3`](https://github.com/UBC-MDS/canadacovidmetrics/commit/341f5f3abd886c0b82dbbee784acbb574d3ac798))
* Version updated ([`d4b8bf0`](https://github.com/UBC-MDS/canadacovidmetrics/commit/d4b8bf028a67ebbc1ceef8f43f5fb2073ce6fd68))
* Typo in datetime check ([`e083ffd`](https://github.com/UBC-MDS/canadacovidmetrics/commit/e083ffdc4164ae85245ba28690c2fe9857d06b53))
* Tests updated to pass ([`c394f8a`](https://github.com/UBC-MDS/canadacovidmetrics/commit/c394f8a6685396b6a8a0aea680c1351a1e7831c0))
* Merge conflict resolved ([`e3aec85`](https://github.com/UBC-MDS/canadacovidmetrics/commit/e3aec8548ba4abfda140d5e842623cbf3bbbe470))
* Ccm alias added to function calls ([`3578d50`](https://github.com/UBC-MDS/canadacovidmetrics/commit/3578d508a875c229e9b5937a63385f92cab3036a))
* Add pytest to package dependencies ([`ac93cca`](https://github.com/UBC-MDS/canadacovidmetrics/commit/ac93cca55e0680e31dc75638e04e8a0b8311ef0d))
* Bug fixes for test_total_cumulative_deaths. Code cleanup ([`044acc4`](https://github.com/UBC-MDS/canadacovidmetrics/commit/044acc45b185786ffe8d0f60f42010c79da97ada))
* Add dependencies and authors to pyproject.toml ([`bd877f2`](https://github.com/UBC-MDS/canadacovidmetrics/commit/bd877f28fdb2c75d197b7883fdb99ea3a73e875b))

### Documentation
* Version bump ([`b1abbd3`](https://github.com/UBC-MDS/canadacovidmetrics/commit/b1abbd3102da74466c6153a87eeb9aa267bd857f))
* Test function docstring updated ([`4959dd6`](https://github.com/UBC-MDS/canadacovidmetrics/commit/4959dd65e247a35839adea74af58efe623a5eaa1))
* Updated default start date in docstring ([`344b25a`](https://github.com/UBC-MDS/canadacovidmetrics/commit/344b25a3bee7df15c83efaf4f5ae2bcde56325de))
* Contributing and conduct files updated ([`7e430cf`](https://github.com/UBC-MDS/canadacovidmetrics/commit/7e430cfb357c134a6f2163f7180866c4598a57f9))

## v0.1.0 (11/01/2022)

- First release of `canadacovidmetrics`!

tests/test_CanadaCovidMetrics.py
from canadacovidmetrics import CanadaCovidMetrics as ccm
import pytest
import pandas as pd

def test_date_format_check():
    """Test date_format_check"""

    # initialise values to test
    date_correct_format = "2021-10-31"
    date_wrong_format = "10-31-2021"
    not_a_date = "121242539"

    # expect None returned for proper date format
    assert ccm.date_format_check(date_correct_format) == None, "Expect None returned for proper date formate failed"
    
    # expect ValueError raised for incorrect date format
    with pytest.raises(ValueError):
        ccm.date_format_check(date_wrong_format)
    
    with pytest.raises(ValueError):
        ccm.date_format_check(not_a_date)

def test_loc_format_check():
    """Test loc_format_check"""

    # initialise values to test
    loc_correct_format = "canada"
    loc_wrong_format = "PEI"
    not_a_loc = "ABC1234!"

    # expect None returned for proper loc format
    assert ccm.loc_format_check(loc_correct_format) == None, "Expect None returned for proper loc formate failed"
    
    # expect ValueError raised for incorrect loc format
    with pytest.raises(ValueError):
        ccm.loc_format_check(loc_wrong_format)
    
    with pytest.raises(ValueError):
        ccm.loc_format_check(not_a_loc)

def test_total_cumulative_recovered_cases():
    """Test loc_format_check"""

    # check default call returns a dataframe object
    assert isinstance(ccm.total_cumulative_recovered_cases(), pd.DataFrame), "Expect dataframe object return failed"

    # check default call returns non-empty dataframe object
    assert ccm.total_cumulative_recovered_cases().empty == False, "Expect non-empty dataframe return failed"
    
    # check default call returns expected dataframe components
    assert (ccm.total_cumulative_recovered_cases().columns == [
        'cumulative_recovered',
        'date_recovered',
        'province',
        'recovered'
        ]).all(), "Expect correct dataframe columns return failed"

    # Test correct datetime type returned if specified
    df = ccm.total_cumulative_recovered_cases(loc='QC', date='2021-05-01', datetype = True)
    assert type(df['date_recovered'][0]) == pd.Timestamp, 'Incorrect datetype returned when datetype is specified true'


def test_total_cumulative_deaths():
    """Test total_cumulative_deaths"""

    # Test correct data is returned when date is specified
    df = ccm.total_cumulative_deaths(loc='BC', date='2021-01-18')
    assert df['cumulative_deaths'].sum() == 1078
    assert df['deaths'].sum() == 31
    
    # Test correct data is returned when date is not specified
    df = ccm.total_cumulative_deaths(loc='BC', date=None, after='2020-01-01', before = '2020-07-30')
    assert df['deaths'].sum() == 194

    # Test correct data size is returned
    df = ccm.total_cumulative_deaths(loc='BC', date=None, after='2020-01-01', before = '2020-10-30')
    assert df.shape == (237, 4)

    # Test correct datetime type returned if specified
    df = ccm.total_cumulative_deaths(loc='QC', date='2021-05-01', datetype = True)
    assert type(df['date_death_report'][0]) == pd.Timestamp, 'Incorrect datetype returned when datetype is specified true'


def test_total_cummulative_cases():
    """ Test total_cummulative_cases"""
    
    # Test correct data is returned when date is specified
    df = ccm.total_cumulative_cases(loc='AB', date='15-03-2020')
    assert df['cumulative_cases'].sum() == 56, 'Incorrect data obtained!'
    assert df['cases'].sum() == 17, 'Incorrect data obtained!'

    # Test correct data is returned when date is not specified
    df = ccm.total_cumulative_cases(loc='AB', date=None, after='2020-01-01', before='25-01-2020')
    assert df['cumulative_cases'].sum() == 0, 'Incorrect data obtained!'
    assert df['cases'].sum() == 0, 'Incorrect data obtained!'
    
    #Test to check if every province data is returned for a given day when loc = "prov"
    df = ccm.total_cumulative_cases(loc='prov', date='15-03-2020')
    assert df["province"].nunique() == 14, 'Incorrect data obtained!'
    
    # Test to check for the correct columns in the dataframe returned
    df = ccm.total_cumulative_cases(loc='AB', date='15-03-2020')
    assert (df.columns == ['cases', 'cumulative_cases', 'date_report', 'province']).all(),  'Incorrect data obtained!'

    # Test correct datetime type returned if specified
    df = ccm.total_cumulative_cases(loc='QC', date='2021-05-01', datetype = True)
    assert type(df['date_report'][0]) == pd.Timestamp, 'Incorrect datetype returned when datetype is specified true'

    
def test_total_cumulative_vaccine_completion():
    """Test total_cumulative_vaccine_completion. Branches inlcude date and no date specified"""

    # Test defaults return a dataframe object
    assert isinstance(ccm.total_cumulative_vaccine_completion(), pd.DataFrame), 'Pandas DataFrame Not Returned'

    # Test default call returns non-empty dataframe object
    assert ccm.total_cumulative_vaccine_completion().empty == False, 'Default returned dataframe is empty'

    # Test defaults returns expected dataframe components
    assert (ccm.total_cumulative_vaccine_completion().columns == [
        'cumulative_cvaccine',
        'cvaccine',
        'date_vaccine_completed',
        'province'
        ]).all(), 'Default returned dataframe has incorrect column names'

    # Test correct data is returned when date is specified
    df = ccm.total_cumulative_vaccine_completion(loc='QC', date='2021-05-01', datetype=False)  
    assert df['cumulative_cvaccine'].sum() == 97108, 'Incorrect cumulative_cvaccine returned when date is specified' 
    assert df['cvaccine'].sum() == 11250, 'Incorrect cvaccine returned when date is specified' 
    print(df['date_vaccine_completed'][0])
    assert df['date_vaccine_completed'][0] == '01-05-2021', 'Incorrect date_vaccine_completed returned when data is specified' 
    assert df['province'][0] == 'Quebec', 'Incorrect province returned when date is specified' 

    # Test correct data is returned when date is not specified
    df = ccm.total_cumulative_vaccine_completion(loc='QC', date=None, after='2021-01-01', before = '2021-07-30', datetype=False)
    assert df['cumulative_cvaccine'].sum() == 151110816, 'Incorrect cumulative_cvaccine returned when date is not specified' 
    assert df['cvaccine'].sum() == 4773367, 'Incorrect cvaccine returned when date is not specified' 
    assert df['date_vaccine_completed'][0] == '12-01-2021', 'Incorrect date_vaccine_completed returned when date is not specified' 
    assert df['province'][0] == 'Quebec', 'Incorrect province returned when date is not specified' 

    # Test correct data size is returned with no date specified
    df = ccm.total_cumulative_vaccine_completion(loc='QC', date=None, after='2021-01-01', before = '2021-07-30')
    assert df.shape == (200, 4), 'Incorrect shape of dataframe returned when date is not specified' 

    # Test correct data size is returned with date specified
    df = ccm.total_cumulative_vaccine_completion(loc='QC', date='2021-05-01')
    assert df.shape == (1, 4), 'Incorrect shape of dataframe returned when date is specified'

    # Test correct datetime type returned if specified
    df = ccm.total_cumulative_vaccine_completion(loc='QC', date='2021-05-01', datetype = True)
    assert type(df['date_vaccine_completed'][0]) == pd.Timestamp, 'Incorrect datetype returned when datetype is specified true'


tests/test_canadacovidmetrics.py
from canadacovidmetrics import canadacovidmetrics as ccm
import pytest
import pandas as pd

def test_date_format_check():
    """Test date_format_check"""

    # initialise values to test
    date_correct_format = "2021-10-31"
    date_wrong_format = "10-31-2021"
    not_a_date = "121242539"

    # expect None returned for proper date format
    assert ccm.date_format_check(date_correct_format) == None, "Expect None returned for proper date formate failed"
    
    # expect ValueError raised for incorrect date format
    with pytest.raises(ValueError):
        ccm.date_format_check(date_wrong_format)
    
    with pytest.raises(ValueError):
        ccm.date_format_check(not_a_date)

def test_loc_format_check():
    """Test loc_format_check"""

    # initialise values to test
    loc_correct_format = "canada"
    loc_wrong_format = "PEI"
    not_a_loc = "ABC1234!"

    # expect None returned for proper loc format
    assert ccm.loc_format_check(loc_correct_format) == None, "Expect None returned for proper loc formate failed"
    
    # expect ValueError raised for incorrect loc format
    with pytest.raises(ValueError):
        ccm.loc_format_check(loc_wrong_format)
    
    with pytest.raises(ValueError):
        ccm.loc_format_check(not_a_loc)

def test_get_recoveries():
    """Test loc_format_check"""

    # check default call returns a dataframe object
    assert isinstance(ccm.get_recoveries(), pd.DataFrame), "Expect dataframe object return failed"

    # check default call returns non-empty dataframe object
    assert ccm.get_recoveries().empty == False, "Expect non-empty dataframe return failed"
    
    # check default call returns expected dataframe components
    assert (ccm.get_recoveries().columns == [
        'cumulative_recovered',
        'date',
        'province',
        'recovered'
        ]).all(), "Expect correct dataframe columns return failed"

    # Test correct datetime type returned if specified
    df = ccm.get_recoveries(loc='QC', date='2021-05-01', datetype = True)
    assert type(df['date'][0]) == pd.Timestamp, 'Incorrect datetype returned when datetype is specified true'


def test_get_deaths():
    """Test get_deaths"""

    # Test correct data is returned when date is specified
    df = ccm.get_deaths(loc='BC', date='2021-01-18')
    assert df['cumulative_deaths'].sum() == 1078
    assert df['deaths'].sum() == 31
    
    # Test correct data is returned when date is not specified
    df = ccm.get_deaths(loc='BC', after='2020-01-01', before = '2020-07-30')
    assert df['deaths'].sum() == 194

    # Test correct data size is returned
    df = ccm.get_deaths(loc='BC', after='2020-01-01', before = '2020-10-30')
    assert df.shape == (237, 4)

    # Test correct datetime type returned if specified
    df = ccm.get_deaths(loc='QC', date='2021-05-01')
    assert type(df['date'][0]) == pd.Timestamp, 'Incorrect datetype returned when datetype is specified true'


def test_get_cases():
    """ Test get_cases"""
    
    # Test correct data is returned when date is specified
    df = ccm.get_cases(loc='AB', date='15-03-2020')
    assert df['cumulative_cases'].sum() == 56, 'Incorrect data obtained!'
    assert df['cases'].sum() == 17, 'Incorrect data obtained!'

    # Test correct data is returned when date is not specified
    df = ccm.get_cases(loc='AB', after='2020-01-01', before='25-01-2020')
    assert df['cumulative_cases'].sum() == 0, 'Incorrect data obtained!'
    assert df['cases'].sum() == 0, 'Incorrect data obtained!'
    
    #Test to check if every province data is returned for a given day when loc = "prov"
    df = ccm.get_cases(loc='prov', date='15-03-2020')
    assert df["province"].nunique() == 14, 'Incorrect data obtained!'
    
    # Test to check for the correct columns in the dataframe returned
    df = ccm.get_cases(loc='AB', date='15-03-2020')
    assert (df.columns == ['cases', 'cumulative_cases', 'date', 'province']).all(),  'Incorrect data obtained!'

    # Test correct datetime type returned if specified
    df = ccm.get_cases(loc='QC', date='2021-05-01')
    assert type(df['date'][0]) == pd.Timestamp, 'Incorrect datetype returned when datetype is specified true'

    
def test_get_vaccinations():
    """Test get_vaccinations. Branches inlcude date and no date specified"""

    # Test defaults return a dataframe object
    assert isinstance(ccm.get_vaccinations(), pd.DataFrame), 'Pandas DataFrame Not Returned'

    # Test default call returns non-empty dataframe object
    assert ccm.get_vaccinations().empty == False, 'Default returned dataframe is empty'

    # Test defaults returns expected dataframe components
    assert (ccm.get_vaccinations().columns == [
        'cumulative_cvaccine',
        'cvaccine',
        'date',
        'province'
        ]).all(), 'Default returned dataframe has incorrect column names'

    # Test correct data is returned when date is specified and datetype is False
    df = ccm.get_vaccinations(loc='QC', date='2021-05-01', datetype=False)  
    assert df['cumulative_cvaccine'].sum() == 97108, 'Incorrect cumulative_cvaccine returned when date is specified' 
    assert df['cvaccine'].sum() == 11250, 'Incorrect cvaccine returned when date is specified' 
    assert df['date'][0] == '01-05-2021', 'Incorrect date returned when data is specified' 
    assert df['province'][0] == 'Quebec', 'Incorrect province returned when date is specified' 

    # Test correct data is returned when date is not specified
    df = ccm.get_vaccinations(loc='QC', after='2021-01-01', before = '2021-07-30', datetype=False)
    assert df['cumulative_cvaccine'].sum() == 151110816, 'Incorrect cumulative_cvaccine returned when date is not specified' 
    assert df['cvaccine'].sum() == 4773367, 'Incorrect cvaccine returned when date is not specified' 
    assert df['date'][0] == '12-01-2021', 'Incorrect date returned when date is not specified' 
    assert df['province'][0] == 'Quebec', 'Incorrect province returned when date is not specified' 

    # Test correct data size is returned with no date specified
    df = ccm.get_vaccinations(loc='QC', after='2021-01-01', before = '2021-07-30')
    assert df.shape == (200, 4), 'Incorrect shape of dataframe returned when date is not specified' 

    # Test correct data size is returned with date specified
    df = ccm.get_vaccinations(loc='QC', date='2021-05-01')
    assert df.shape == (1, 4), 'Incorrect shape of dataframe returned when date is specified'

    # Test correct datetime type returned if specified
    df = ccm.get_vaccinations(loc='QC', date='2021-05-01')
    assert type(df['date'][0]) == pd.Timestamp, 'Incorrect datetype returned when datetype is specified true'

src/CanadaCovidMetrics/CanadaCovidMetrics.py
import datetime as dt
from multiprocessing.sharedctypes import Value
import requests
import pandas as pd

today = '{}'.format(dt.date.today())

def date_format_check(datestr):
    """Check date format is compatible with API call
    Parameters
    ----------
    datestr : str
        Date string to be checked
    Returns
    -------
    None
    Examples
    --------
    >>> date_format_check('2021-03-30')
    """
    format_ymd = "%Y-%m-%d"
    format_dmy = "%d-%m-%Y"
    try:
        dt.datetime.strptime(datestr, format_ymd)
    except:
        try:
            dt.datetime.strptime(datestr, format_dmy)
        except:
            raise(ValueError("Incorrect date format {} : date format should be either YYYY-MM-DD or DD-MM-YYYY".format(datestr)))
    
    return


def loc_format_check(locstr):
    """Check province/location format is compatible with API call
    
    Parameters
    ----------
    locstr : str
        Location string to be checked
    Returns
    -------
    None
    Examples
    --------
    >>> loc_format_check('2021-03-30')
    """
    format_loc = ['canada', 'prov', 'BC', 'AB', 'SK', 'MB', 'ON', 'QC', 'NL', 'NB', 'NS', 'PE', 'NT', 'YT', 'NU', 'RP']
    if locstr not in format_loc:
        raise(ValueError("Value passed for loc argument is not recognized. Must be one of: 'prov', 'canada', or a two-letter capitalized province code"))

    return

def total_cumulative_cases(loc='prov', date=None, after='2020-01-01', before=today, datetype=True):
    """Query total cumulative cases with ability to specify \
        location and date range of returned data.
    Parameters
    ----------
    loc : str
        Specify geographic filter and aggregation of returned data.
        Valid loc arguments are: 'canada', 'prov' and two-letter \
        province codes (e.g. 'ON', 'BC', etc.)
    date : str
        If not None, return data from the specified date YYYY-MM-DD.
        Superceeds 'after' and 'before' parameters.
    after : str
        Return data on and after the specified date YYYY-MM-DD.
    before : str
        Return data on and before the specified date YYYY-MM-DD.
    datetype : boolean
        Return date column as a string (False) or as a datetime (True).
        Default is True.
    Returns
    -------
    dict
        JSON response from queried api.
    Examples
    --------
    >>> total_cumulative_cases(loc='BC')
    """
     
    loc_format_check(loc)  # check location is valid

    if date is not None:
        date_format_check(date)  # check date is valid
        url = 'https://api.opencovid.ca/timeseries?stat=cases&loc={}&date={}'.format(loc, date) 
    else:
        date_format_check(before)  # check before-date is valid
        date_format_check(after)  # check after-date is valid
        url = 'https://api.opencovid.ca/timeseries?stat=cases&loc={}&after={}&before={}'.format(loc, after, before)
    
    r = requests.get(url = url)
    json_body = r.json()['cases']
    df = pd.DataFrame.from_dict(json_body)

    if datetype:
        df['date_report']= df['date_report'].apply(lambda x: dt.datetime.strptime(x, '%d-%m-%Y'))
    
    return df


def total_cumulative_deaths(loc='prov', date=None, after='2020-01-01', before=today, datetype=True):
    """Query total cumulative deaths with ability to specify \
        location and date range of returned data.
    Parameters
    ----------
    loc : str
        Specify geographic filter and aggregation of returned data.
        Valid loc arguments are: 'canada', 'prov' and two-letter \
        province codes (e.g. 'ON', 'BC', etc.)
    date : str
        If not None, return data from the specified date YYYY-MM-DD.
        Superceeds 'after' and 'before' parameters.
    after : str
        Return data on and after the specified date YYYY-MM-DD.
    before : str
        Return data on and before the specified date YYYY-MM-DD.
    datetype : boolean
        Return date column as a string (False) or as a datetime (True).
        Default is True.
    Returns
    -------
    df
        Pandas dataframe containing content of API response.
    Examples
    --------
    >>> total_cumulative_deaths(loc='BC')
    """  

    loc_format_check(loc)  # check location is valid

    if date is not None:
        date_format_check(date)  # check date is valid
        url = 'https://api.opencovid.ca/timeseries?stat=mortality&loc={}&date={}'.format(loc, date) 
    else:
        date_format_check(before)  # check before-date is valid
        date_format_check(after)  # check after-date is valid
        url = 'https://api.opencovid.ca/timeseries?stat=mortality&loc={}&after={}&before={}'.format(loc, after, before)
    
    r = requests.get(url = url)
    json_body = r.json()['mortality']
    df = pd.DataFrame.from_dict(json_body)

    if datetype:   
        df['date_death_report']= df['date_death_report'].apply(lambda x: dt.datetime.strptime(x, '%d-%m-%Y'))

    return df


def total_cumulative_recovered_cases(loc='prov', date=None, after='2020-01-01', before=today, datetype=True):
    """Query total cumulative recovered cases with ability \
        to specify location and date range of returned data.
    Parameters
    ----------
    loc : str
        Specify geographic filter and aggregation of returned data.
        Valid loc arguments are: 'canada', 'prov' and two-letter \
        province codes (e.g. 'ON', 'BC', etc.)
    date : str
        If not None, return data from the specified date YYYY-MM-DD.
        Superceeds 'after' and 'before' parameters.
    after : str
        Return data on and after the specified date YYYY-MM-DD.
    before : str
        Return data on and before the specified date YYYY-MM-DD.
    datetype : boolean
        Return date column as a string (False) or as a datetime (True).
        Default is True.
    Returns
    -------
    df
        Pandas dataframe containing content of API response.
    Examples
    --------
    >>> total_cumulative_recovered_cases(loc='BC')
    """  
    
    loc_format_check(loc)  # check location is valid

    if date is not None:
        date_format_check(date)  # check date is valid
        url = 'https://api.opencovid.ca/timeseries?stat=recovered&loc={}&date={}'.format(loc, date) 
    else:
        date_format_check(before)  # check before-date is valid
        date_format_check(after)  # check after-date is valid
        url = 'https://api.opencovid.ca/timeseries?stat=recovered&loc={}&after={}&before={}'.format(loc, after, before)
    
    r = requests.get(url = url)
    json_body = r.json()['recovered']
    df = pd.DataFrame.from_dict(json_body)

    if datetype:
        df['date_recovered']= df['date_recovered'].apply(lambda x: dt.datetime.strptime(x, '%d-%m-%Y'))

    return df


def total_cumulative_vaccine_completion(loc='prov', date=None, after='2021-01-01', before=today, datetype=True):
    """Query total cumulative vaccine completion with ability \
        to specify location and date range of returned data.

    Parameters
    ----------
    loc : str
        Specify geographic filter and aggregation of returned data.
        Valid loc arguments are: 'canada', 'prov' and two-letter \
        province codes (e.g. 'ON', 'BC', etc.)
    date : str
        If not None, return data from the specified date YYYY-MM-DD.
        Superceeds 'after' and 'before' parameters.
    after : str
        Return data on and after the specified date YYYY-MM-DD.
        Default is 2021-01-01.
    before : str
        Return data on and before the specified date YYYY-MM-DD.
        Default is the date of query.
    datetype : boolean
        Return date column as a string (False) or as a datetime (True).
        Default is True.
    Returns
    -------
    df
        Pandas dataframe containing content of API response.
    Examples
    --------
    >>> total_cumulative_vaccine_completion(loc='BC')
    """

    loc_format_check(loc)  # check location is valid

    if date is not None:
        date_format_check(date)  # check date is valid
        url = 'https://api.opencovid.ca/timeseries?stat=cvaccine&loc={}&date={}'.format(loc, date) 
    else:
        date_format_check(before)  # check before-date is valid
        date_format_check(after)  # check after-date is valid
        url = 'https://api.opencovid.ca/timeseries?stat=cvaccine&loc={}&after={}&before={}'.format(loc, after, before)
    
    r = requests.get(url = url)
    json_body = r.json()['cvaccine']
    df = pd.DataFrame.from_dict(json_body)

    if datetype:
        df['date_vaccine_completed']= df['date_vaccine_completed'].apply(lambda x: dt.datetime.strptime(x, '%d-%m-%Y'))

    return df